src/interpolate.rs

```diff
@@ -0,0 +1,249 @@
+use {Duplex, Frame, Sample};
+
+use core::marker::PhantomData;
+
+/// An iterator that converts the rate at which frames are yielded from some given frame
+/// Interpolator into a new type.
+///
+/// Other names for `sample::rate::Converter` might include:
+///
+/// - Sample rate converter
+/// - {Up/Down}sampler
+/// - Sample interpolater
+/// - Sample decimator
+///
+#[derive(Clone)]
+pub struct Converter<T: Iterator, I: Interpolator<T>>
+    where <T as Iterator>::Item: Frame
+{
+    interpolator: I,
+    interpolation_value: f64,
+    source_to_target_ratio: f64,
+    phantom: PhantomData<T>
+}
+
+/// Interpolator that just rounds off any values to the previous value from the source
+pub struct Floor<I>
+    where I: Iterator,
+          I::Item: Frame
+{
+    source: I,
+    left: I::Item 
+}
+
+/// Interpolator that interpolates linearly between the previous value and the next value
+pub struct Linear<I>
+    where I: Iterator,
+          I::Item: Frame
+{
+    source: I,
+    left: I::Item,
+    right: Option<I::Item>
+}
+
+/// Trait for all things that can interpolate between two values. Implementations should keep track
+/// of the necessary data both before and after the current frame.
+pub trait Interpolator<I>
+    where I: Iterator,
+          I::Item: Frame
+{
+    /// Creates a new Interpolator using the provided Iterator as the source
+    fn from_source(source: I) -> Self;
+
+    fn source(&self) -> &I;
```

Comment:
It seems like the `source` or `from_source` methods don't really need to be a part of the trait here? They don't have much to do with interpolation, and would probably be better as inherent methods on the types that require them.